File: supreme/supreme.go
package supreme

import (
	"net/http"
	"net/http/cookiejar"
	"net/url"
	"strconv"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"golang.org/x/net/publicsuffix"
)

type Client struct {
	http.Client
}

func NewClient() (*Client, error) {
	jar, err := cookiejar.New(&cookiejar.Options{PublicSuffixList: publicsuffix.List})
	if err != nil {
		return nil, err
	}

	return &Client{http.Client{Jar: jar}}, nil
}

func setUserAgent(req *http.Request) {
	req.Header.Set("User-Agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/63.0.3239.84 Safari/537.36")
}

func (c *Client) AddItem() error {
	u := "http://www.supremenewyork.com/shop/171001/add"
	v := url.Values{}
	v.Set("utf8", "✓")
	v.Set("st", "18260")
	v.Set("s", "49858")
	v.Set("commit", "add to cart")

	req, err := http.NewRequest("POST", u, strings.NewReader(v.Encode()))
	if err != nil {
		return err
	}

	setUserAgent(req)
	resp, err := c.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	return nil
}

func (c *Client) NumItems() (int, error) {
	u := "http://www.supremenewyork.com/shop/cart"
	req, err := http.NewRequest("GET", u, nil)
	if err != nil {
		return 0, err
	}

	setUserAgent(req)
	resp, err := c.Do(req)
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()

	doc, err := goquery.NewDocumentFromResponse(resp)
	if err != nil {
		return 0, err
	}

	numItems, err := strconv.Atoi(strings.Split(doc.Find("#items-count").First().Text(), " ")[0])
	if err != nil {
		return 0, err
	}

	return numItems, nil
}

File: src/supreme.js
const rp = require('request-promise');

const Clothing = require('./clothing');

const baseUrl = 'http://www.supremenewyork.com';
const userAgent = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/63.0.3239.84 Safari/537.36';

class Supreme {
  constructor() {
    this._jar = rp.jar();
    this._request = rp.defaults({
      followAllRedirects: true,
      headers: {
        'User-Agent': userAgent,
      },
      jar: this._jar,
      resolveWithFullResponse: true,
    });
  }

  async addPendingClothing(pendingClothing) {
    for (let { uri, st, s } of pendingClothing) {
      await this._request({
        method: 'POST',
        uri: `${baseUrl}${uri}`,
        formData: {
          utf8: '✓',
          st,
          s,
          commit: 'add to cart',
        },
      });
    }
  }

  async fetchNewClothingHtmlDocs() {
    try {
      const res = await this._request({
        method: 'GET',
        uri: 'http://www.supremenewyork.com/shop/new',
      });
      const $ = cheerio.load(res.body);
      const requests = $('div.inner-article > a')
        .toArray()
        .map(elem => elem.attribs.href)
        .filter(url => url.match(Clothing.urlRegex))
        .map(url => this._request({ method: 'GET', uri: `${baseUrl}${url}`}));
      await Promise.all(requests);
      return requests.map(r => r.response.body);
    } catch (err) {
      throw err;
    }
  }

  getSession() {
    const { cookies } = this._jar._jar.toJSON();
    return cookies.find(({ key }) => key === '_supreme_sess' );
  }
}

module.exports = Supreme;
